src/core/gui/XournalppCursor.cpp

```diff
@@ -515,6 +520,68 @@ auto XournalppCursor::createHighlighterOrPenCursor(int size, double alpha) -> Gd
 }
 
 
+auto XournalppCursor::createHorizontalLineCursor() -> GdkCursor* {
+    GtkWidget* theWidget = control->getWindow()->getXournal()->getWidget();
+    GdkWindow* theWindow = gtk_widget_get_window(theWidget);
+
+    gint x, y;
+    GdkDevice *mouse_device;
+
+    #if GTK_CHECK_VERSION (3,20,0)
+    GdkSeat *seat = gdk_display_get_default_seat (gdk_display_get_default ());
+    mouse_device = gdk_seat_get_pointer (seat);
+    #else
+    GdkDeviceManager *devman = gdk_display_get_device_manager (gdk_display_get_default ());
+    mouse_device = gdk_device_manager_get_client_pointer (devman);
+    #endif
+
+    gdk_window_get_device_position (theWindow, mouse_device, &x, &y, nullptr);
+
+    Layout* const layout = this->control->getWindow()->getLayout();
+    XojPageView* page = layout->getPageViewAt(x, y);
+    if (page) {
+        Range range = page->getVisiblePart();
+        double zoom = control->getZoomControl()->getZoom();
+        int width = static_cast<int>(std::round(range.getWidth() * zoom));
+        int height = 5; // Five is the default cursor size used elsewhere in the app.
+        gint dx = x - static_cast<int>(range.getX() * zoom) - page->getX();
+        if (this->currentCursorFlavour == (dx << (sizeof(gint) * 4) | width)) {
+            return nullptr;
+        }
+
+        cairo_surface_t* crCursor = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, width, height);
+        cairo_t* cr = cairo_create(crCursor);
+
+        cairo_set_line_width(cr, 1.2);
+        Util::cairo_set_source_rgbi(cr, this->control->getSettings()->getSelectionColor());
+        cairo_move_to(cr, 0, 0);
+        cairo_line_to(cr, width, 0);
+        cairo_close_path(cr);
+        cairo_stroke(cr);
+        cairo_destroy(cr);
+        GdkPixbuf* pixbuf = xoj_pixbuf_get_from_surface(crCursor, 0, 0, width, height);
+        cairo_surface_destroy(crCursor);
+        GdkCursor* gdkCursor = gdk_cursor_new_from_pixbuf(
+                gtk_widget_get_display(theWidget), pixbuf, dx, 0);
+        g_object_unref(pixbuf);
+        this->currentCursor = CRSR_HORIZONTALLINE;
+        // Shift the 'dx' and 'width' variables together to check if we need
+        // to re-draw the cursor.
+        this->currentCursorFlavour = (gulong) (dx << (sizeof(gint) * 4) | width);
```

Comment:
This is hard to read and error prone. It would be better to define a union like this:
```c++
using HorizontalLineFlavour = union {
  struct {
    int dx : 32;
    int width : 32;
  } data;
  gulong raw;
};
static_assert(sizeof(HorizontalLineFlavour{}.data) == sizeof(HorizontalLineFlavour{}.raw),
              "Size of packed representation of HorizontalLineFlavour does not match size of integer representation");

// retrieve:
gulong myRawFlavour = this->currentCursorFlavour.raw;

// set:
this->currentCursorFlavour = HorizontalLineFlavour{data, width}.raw;
```

Reply:
Wow, that's pretty slick.

Reply:
FWIW per https://en.cppreference.com/w/cpp/language/union, "It is undefined behavior to read from the member of the union that wasn't most recently written. Many compilers implement, as a non-standard language extension, the ability to read inactive members of a union."

The set example above of `HorizontalLineFlavour{data, width}.raw` is therefore undefined (the initializer writes to the `data` member, while the subsequent access reads from the `.raw` member (also I think you meant `HorizontalLineFlavour{dx, width}.raw`).

I see the `static_assert` making sure the sizes are the same, but that doesn't really address the UB.

FTR here's a compilable version of that example which, unfortunately, produces no warnings despite the UB. https://godbolt.org/z/j7dGcszqn

Reply:
I didn't know that was UB... I think it's probably fine to do the bitshifting, as long as it's well documented in comments. That seems to be the common pattern in this file anyway (and some of these other ones look a lot more complicated with a whole lot less explanation.

Reply:
This is a common and legal pattern in C, so this is disappointing to learn about.